onnxruntime/core/providers/cpu/tensor/concat.cc

```diff
@@ -35,66 +35,116 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx,
 
   // Ensure all of the non concatenated axes match each other
-  for (int index = 1; index < input_count; index++) {
-    tensor_pointer = input_tensors[index];
-    if (tensor_pointer == nullptr)
-      return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
-    auto& inputs_n = *tensor_pointer;
-    const auto& inputs_n_dims = inputs_n.Shape().GetDims();
-    const size_t inputs_n_rank = inputs_n_dims.size();
-    ORT_ENFORCE(inputs_n_rank == inputs_0_rank,
-                "Ranks of input data are different, cannot concatenate them. expected rank: ",
-                inputs_0_rank, " got: ", inputs_n_rank);
-    // Ensure all the other (non-concat) axes match
-    int64_t tensor_size = 1;
-    for (size_t axis_index = 0; axis_index < inputs_0_rank; ++axis_index) {
-      auto dim_value = inputs_n_dims[axis_index];
-      tensor_size *= dim_value;
-
-      // In 'concat' mode, the axis to be concatenated may be different
-      // But in 'stack' mode, all input shapes must be the same and must be validated
-      if (!is_stack_ && axis_index == p.axis)
-        continue;
-
-      ORT_RETURN_IF_NOT(dim_value == inputs_0_dims[axis_index],
-                        "Non concat axis dimensions must match: Axis ",
-                        axis_index, " has mismatched dimensions of ", dim_value,
-                        " and ", inputs_0_dims[axis_index]);
-    }
+  for (int index = reference_tensor_index + 1; index < input_count; index++) {
+    const auto* input = input_tensors[index];
+    ORT_ENFORCE(input != nullptr, "input count mismatch");
+    const auto& input_shape = input->Shape();
+    const auto& input_dims = input_shape.GetDims();
+
+    // Skip shape/rank validation for inputs that are empty.
+    // The ONNX spec states that all dim values along axes not concatentated on
+    // need to be the same for all inputs (empty inputs are not explicitly exempted).
+    // The model in GH issue 8020 has a bunch of Loop nodes all feeding into
+    // the 'Concat' node and one of these Loops tend to have an iteration
+    // count of 0 for some inputs. If the iteration count for a Loop is zero,
+    // we don't execute its subgraph (since the outputs are going to be empty anyway)
+    // and we send an "empty" tensor(s) downstream and use ONNX shape inferred shape
+    // to "compose" the shape for these empty tensor(s).
+    // If we encounter symbolic dims in the ONNX shape inferred shape, we place a '0'
+    // in that position and due to the "lossy" nature of this process, the inputs' shape
+    // validation for such empty inputs fail and hence we skip these validations for all
+    // empty inputs.
+    // This isn't too bad as we will never use empty inputs while concatenating anyway.
+    // We just loosen this check to unblock model in GH issue 8020 to complete processing.
+    if (input_shape.Size() == 0) {
+      input_tensor_sizes.push_back(0);
+    } else {
+      const size_t input_rank = input_dims.size();
```

Comment:
side changes: previously we treat the very first input as the "reference tensor" to perform shape/rank validations on other input tensors. Slightly adjusted the logic such that the reference tensor is the first non-empty tensor (as empty tensors have no role to play in the validation anyway)